streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -496,23 +514,68 @@ private void addPartitions(Collection<TopicPartition> assignment) {
         lastClean = time.milliseconds();
     }
 
-    private void removePartitions() {
-
+    private void removeStreamTasks() {
         // TODO: change this clearing tasks behavior
-        for (StreamTask task : tasks.values()) {
-            log.info("Removing task {}", task.id());
-            try {
-                task.close();
-            } catch (Exception e) {
-                log.error("Failed to close a task #" + task.id() + " in thread [" + this.getName() + "]: ", e);
-                throw e;
-            }
-            sensors.taskDestructionSensor.record();
+        for (StreamTask task : activeTasks.values()) {
+            closeOne(task);
         }
+
         prevTasks.clear();
-        prevTasks.addAll(tasks.keySet());
+        prevTasks.addAll(activeTasks.keySet());
+
+        activeTasks.clear();
+    }
+
+    private void closeOne(AbstractTask task) {
+        log.info("Removing a task {}", task.id());
+        try {
+            task.close();
+        } catch (Exception e) {
+            log.error("Failed to close a " + task.getClass().getSimpleName() + " #" + task.id() + " in thread [" + this.getName() + "]: ", e);
+            throw e;
+        }
+        sensors.taskDestructionSensor.record();
+    }
+
+    protected StandbyTask createStandbyTask(TaskId id) {
+        sensors.taskCreationSensor.record();
+
+        ProcessorTopology topology = builder.build(id.topicGroupId);
+
+        return new StandbyTask(id, restoreConsumer, topology, config, sensors);
+    }
+
+    private void addStandbyTasks() {
+        Map<TopicPartition, Long> checkpointedOffsets = new HashMap<>();
+
+        for (TaskId taskId : partitionGrouper.standbyTasks()) {
+            StandbyTask task = createStandbyTask(taskId);
+            standbyTasks.put(taskId, task);
+            checkpointedOffsets.putAll(task.checkpointedOffsets());
```

Comment:
You are right, this is fine. I was confused about the checkpointedOffsets in standbyTasks before.